[Merged by Bors] - add `Res::clone`

BoxyUwU opened the pull request:
# Objective
Make `Res` cloneable
## Solution
Add an associated fn `clone(self: &Self) -. Self` instead of `Copy + Clone` trait impls to avoid `res.clone()` failing to clone out the underlying `T`

MrGVSV commented:
Would this prevent the contained data from being cloned by deref?

```
let my_cloned_data = res.clone();
```

BoxyUwU commented:
the alternative form i think is actually `(&*res).clone()` but yes

james7132 commented:
Not a fan of this change. The expected behavior of a resource that is clonable should give you a cloned version of the inner value. If we explicitly want to create a new `Res<T>` from another one, I would much rather be explicit about it and use `Res::copy` or something similar (akin to `Arc::downgrade` and other similar smart pointer APIs).

Does calling `res.clone()` on a `Res<T: !Clone>` give you `&T` due to Deref?

bjorn3 commented on crates/bevy_ecs/src/system/system_param.rs:
```diff
@@ -207,6 +207,13 @@ pub struct Res<'w, T: Resource> {
 
+impl<'w, T: Resource> Copy for Res<'w, T> {}
+impl<'w, T: Resource> Clone for Res<'w, T> {
+    fn clone(&self) -> Self {
```
Maybe add `#[inline]`?

mcobzarenco commented:
>Not a fan of this change. 

Same here

> The expected behavior of a resource that is clonable should give you a cloned version of the inner value. If we explicitly want to create a new Res<T> from another one, I would much rather be explicit about it and use Res::copy(res) or something similar (akin to Arc::downgrade and other similar smart pointer APIs that avoid using self to ensure that it's not in conflict with an deref'ed function).

Indeed, this is what `std` does, the `Clone` impls for smart pointers (`Rc`, `Arc` etc.) clones the underlying resource via deref and you're [expected](https://doc.rust-lang.org/std/rc/struct.Rc.html#method.clone) to use e.g. `Arc::clone` to clone the pointer.

james7132 commented:
> Indeed, this is what `std` does, the `Clone` impls for smart pointers (`Rc`, `Arc` etc.) clones the underlying resource via deref and you're [expected](https://doc.rust-lang.org/std/rc/struct.Rc.html#method.clone) to use e.g. `Arc::clone` to clone the pointer.

Actually both `Rc` and `Arc` clone the pointer, only `Box` clones both the pointer and value. Still, I expect the base `Res<T>` pointer to act like `&T`, and ultimately that's what enables the ergonomics around it.

BoxyUwU commented:
`&&T` also has the issue of `.clone()` not cloning the inner `T` so this is not actually limited to `Res<T>: Copy + Clone` but regardless I removed the trait impls and just added `Res::clone` since we ought to provide _some_ way of doing this even if the obvious way of `Copy` is disliked

alice-i-cecile commented on crates/bevy_ecs/src/system/system_param.rs:
```diff
@@ -220,6 +220,17 @@ where
 impl<'w, T: Resource> Res<'w, T> {
+    // no it shouldn't clippy
+    #[allow(clippy::should_implement_trait)]
+    pub fn clone(this: &Self) -> Self {
```
This should probably be `#[inline]` as a tiny method.

alice-i-cecile commented:
After reading the discussion, I think I prefer the current design over implementing the trait.

cart commented on crates/bevy_ecs/src/system/system_param.rs:
```diff
@@ -252,6 +252,17 @@ where
 }
 
 impl<'w, T: Resource> Res<'w, T> {
```
Just throwing an idea out there: maybe we should call this something else to avoid the ambiguity entirely?
(ex: `duplicate()`)

alice-i-cecile replied:
Yep, that's my preference here too.

cart replied:
Hmm actually within the context of requiring `Res::clone` to resolve the ambiguity, I think thats nice and clear, given that we're cloning the Res. I think we should merge this as-is.

cart commented:
bors r+
